src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -281,14 +288,29 @@ public ResultMessage.Rows execute(QueryState state, QueryOptions options, long q
 
         Selectors selectors = selection.newSelectors(options);
         AggregationSpecification aggregationSpec = getAggregationSpec(options);
-        ReadQuery query = getQuery(options, state.getClientState(), selectors.getColumnFilter(),
-                                   nowInSec, userLimit, userPerPartitionLimit, pageSize, aggregationSpec);
+        DataLimits limit = getDataLimits(userLimit, userPerPartitionLimit, pageSize, aggregationSpec);
+
+        // Handle additional validation for topK queries
+        if (restrictions.isTopK())
+        {
+            checkFalse(aggregationSpec != null, TOPK_AGGREGATION_ERROR);
+
+            checkFalse(options.getConsistency() != ConsistencyLevel.ONE &&
```

Comment:
Rather than failing queries not using LOCAL_ONE/ONE I think we should downgrade the CL to ONE/LOCAL_ONE (depending on if the incoming query used LOCAL_) and set a Client Warn that it happened.
It will be very painful for end users if they have tospecially manage that ANN read queries use ONE when everything else in their app is using LOCAL_QUORUM...

Reply:
As discussed on Slack conversation, we should do the same with the check for page size right below. So a page size below the query limit will produce a client warning and be ignored.